In [4]:
#Consider the string given below and answer the following questions 
string = "OQYWFClFhFGAvIWYwGKpmZhnJiyzTslSIhSwvOsqJMEphzmifTkyqOMNpnOtXZxmCfgDYqbaBHAUvIWhMnvwZnEMVDvmEfLrDoQnAZgQEgXQVnmSYkfedpAdhrtpOgORpYLRZYGWdhWYuqQssCUXtTzKRDAhpjUheOzUroZNzWFtZOVwIapzUYtbSbjYNErzQ"
print(len(string))

193


####  Calculate the number of characters in the string

-  193
-  200
-  300
-  500


In [5]:
# Type your code here
cnt=string.count('a')
print(cnt)

2


#### Count the number of occurences of 'a' in the string

-  1
-  2
-  10
-  5

In [7]:
# Type your code here
print(string.startswith('if'))

False


#### Identify whether the string starts with 'if'.

-  True
-  False

In [8]:
# Type your code here
print(string[63:88])

mCfgDYqbaBHAUvIWhMnvwZnEM


#### Return the substring starting at index 63 and ending at index 88  [Both the indexes are inclusive]

-  'OQYWFClFhFGA'
-  'mCfgDYqbaBHAUvIWhMnvwZnEMV'
-  'IWYwGKpmZhnJiyzTslSIhSwv'
-  'JMEphzmifTkyqOMNpnOtXZ'

In [9]:
# Type your code here
print(string[45])

z


#### Identify the character at index 45

-  a
-  b
-  z
-  x

In [ ]:
# Type your code here

In [6]:
A = [5,1,3,4,4,5,6,7]

B = [3,3,5,5, 1 ,7 ,2]

set1=set(A)
set2=set(B)
print(set1)
print(set2)
answer=(set1 | set2)
print(answer)

{1, 3, 4, 5, 6, 7}
{1, 2, 3, 5, 7}
{1, 2, 3, 4, 5, 6, 7}


In [9]:

#d = dict(a=1, b=2)

#d = {'a':1, 'b':2}
d = {}
d


{}

In [1]:
func(x,y):
    z = x*y + x + y

func(2)

SyntaxError: invalid syntax (<ipython-input-1-b35086486212>, line 1)

In [2]:
def func(x,y =1):
    z = x*y + x + y
    return z

func(2,func(3))

23

In [4]:
def func(x,y):
    z = x*y + x + y
    return z

func(2,func(3))

TypeError: func() missing 1 required positional argument: 'y'

In [5]:
def func(x=1 ,y=2):
    z = x*y + x + y
    return z

func(2,func(3))

35

In [6]:
def say(message, times = 1):
    print(message * times)

say('Hello')
say('World', 5)

Hello
WorldWorldWorldWorldWorld


In [7]:
min = (lambda x, y: x if x < y else y)
min(101*99, 102*98)

9996

In [8]:
list_numbers = (1,2,3,4)
sample_map = map(lambda x: x*2, list_numbers)
print(list(sample_map))

[2, 4, 6, 8]


In [9]:
def multi(x):
    return x*2
list_numbers = [1,2,3,4]
sample_map = map(multi, list_numbers)
print(list(sample_map))

[2, 4, 6, 8]


In [11]:
class Student:
    standard = 'IV'
    def __init__(self,age,name,Cid):
        self.age = age
        self.name = name
        self.id = Cid

In [12]:
student1 = Student(12,"Raj",24)

In [13]:
student1 = Student("Shiv",4, 3)

In [14]:
class A :
    x = 10
    def __init__(self, y,z):
        self.y = y
        self.z = z
           
    def update_y(self):
        self.y = self.y * self.x
        self.z = self.z * self.x
        
A1 = A(3,4)
A2 = A(5,6)

A1.update_y()
A1.y + A2.z

36

In [15]:
class A :
    x = 10
    def __init__(self, y,z):
        self.y = y
        self.z = z
           
    def update_y(self):
        self.y = self.y * self.x
   
    @classmethod
    def update_z(cls):
        self.z = self.z + 20

        
A1 = A(3,4)
A2 = A(5,6)

A2.update_z()
A1.y + A2.z

NameError: name 'self' is not defined

In [16]:
class A :
    x = 10
    def __init__(self, y,z):
        self.y = y
        self.z = z
           
    def update_y(self):
        self.y = self.y * self.z

        
        
A1 = A(3,4)
A2 = A(5,6)

A.x = 30
A1.y + A2.x

33

In [17]:
class A:
    def __init__(self, x=1):
        self.x = x
class B(A):
    def __init__(self,y =2):
        super().__init__()
        self.y = y

def main():
    b = B()
    print(b.x,b.y)
main()

1 2


In [25]:
import ast,sys
input_str = "bmaunmdbraai"

message1= input_str[::2]
message2 = input_str[1::2]
print(message1+','+message2)
 

bandra,mumbai


In [26]:
d = 23//5
print(d)

4


In [36]:

if (1)

    print (True)

SyntaxError: invalid syntax (<ipython-input-36-39f452702353>, line 2)

In [37]:
for i in range(3):
    print(i)

0
1
2


In [46]:
a=11

b=12

c=13 

d=14

print(f'{a}{b}{c}')

print(d)

111213
14


In [43]:
L=[2, 3, 4, 2, 1, 2, 3]
print(L.index(2))

0


In [48]:
L = [2, 1, 2, 4, 5, 3, 6]
K = 4 in L
K

True

In [56]:
L=[10,20,30,40,50,60,70,80,90,100]
L[2:-1:2]

[L[i] for i in range(1,9,2)]

[20, 40, 60, 80]

In [60]:
A = [10, 20, 30] 
B = [45, 35, 25]
B.extend(A)
B

[45, 35, 25, 10, 20, 30]

In [1]:
print(chr(51))

3


In [7]:
Str=" djsfkj dfnjkdnf "
Str.strip()
S

'djsfkj dfnjkdnf'

In [12]:

s = "upgrad"

s = s[:2] + s[2].upper() + s[3:]

In [14]:
s='aabupGradaab'.strip('a').strip('b').strip('a')
s

'upGrad'

In [16]:

s='aabupGradaab'.strip('aab')
s

'upGrad'

In [20]:
'G'.join(['up','rad'])

'upGrad'

In [18]:

s = upgrad

s = s[:2] + s[2].upper() + s[3:]

NameError: name 'upgrad' is not defined

In [19]:
s = upGrad

NameError: name 'upGrad' is not defined

In [21]:
txt = "Hello World" [::-1] 
print(txt)

dlroW olleH


In [22]:
d={'a':64, 'b':65, 'c':66, 'd':67}
print(d['e'])

KeyError: 'e'

In [23]:
d={'a':64, 'b':65, 'c':66, 'd':67}
print(d.keys())

dict_keys(['a', 'b', 'c', 'd'])


In [32]:
dict_1 = {"Python'':40, "R'':45}
dict_1.delete('R':45

SyntaxError: invalid syntax (<ipython-input-32-11b60d6b8ee7>, line 2)

In [33]:
L1=[2, 3, 4, 2, 1, 2, 3]
L2=[2, 3, 4, 2, 1, 2, 3]

L1+L2


[2, 3, 4, 2, 1, 2, 3, 2, 3, 4, 2, 1, 2, 3]

In [39]:
import numpy as np

np.array( [1.0, 2, 3.5, 0.2, True] )


array([1. , 2. , 3.5, 0.2, 1. ])

In [42]:
array_1 = [1, 2, 3, 5, 4, 6, 7, 8, 5, 3, 2]
print(array_1[2:-3])

[3, 5, 4, 6, 7, 8]


In [48]:
array_1 = [1, 2, 3, 5, 4, 6, 7, 8, 5, 3, 2]

#print(array_1[ : : -2])

print(array_1[1: : 2])



[2, 5, 6, 8, 3]


In [52]:
list_1 = [48.49, 67.54, 57.47, 68.17, 51.18, 68.31, 50.33, 66.7, 45.62, 43.59, 53.64, 70.08, 47.69, 61.27, 44.14, 51.62, 48.72, 65.11]
np.percentile(list_1,25)
np.percentile(list_1,75)

66.30250000000001

In [53]:
variable_list = [5, "string", 1.4]

for number in variable_list:
    try:
        range_list = list(range(number))
        print(number, "-", range_list)

    except:
        print(number, "- not an integer")

5 - [0, 1, 2, 3, 4]
string - not an integer
1.4 - not an integer


In [54]:
class GameProtagonist:
    def __init__(self):
        self.health = 100
        self.mana = 100
        self.level = 1

    def take_damage(self):
        self.health -= 10

player_one = GameProtagonist()
player_one.health

100

In [59]:
new_list = [num for num in original_list if num % 2 == 0]
new_list

[10, 12, 4, 6]

In [60]:
original_list = [10, 12, 17, 29, 4, 6]
new_list = []

for num in original_list:
    if num % 2 == 0:
        new_list.append(num)
new_list

[10, 12, 4, 6]

In [61]:
import re

string = "This is a string"
re.findall('is', string)

['is', 'is']

In [62]:
def multiples_of(number):
    multiples = []
    for i in range(1, 6):
        multiples.append(i * number)
    return multiples

In [63]:
multiples_of(number)

[1.4, 2.8, 4.199999999999999, 5.6, 7.0]

In [64]:
def multiples_of(number):
    for i in range(1, 6):
        return i * number
multiples_of(number)

1.4

In [65]:
def multiples_of(number):
    multiples = []
    for i in range(1, 6):
        multiples.append(i * number)
    yield multiples
multiples_of(number)

<generator object multiples_of at 0x00000276E46D3ED0>

In [69]:
class Circle:
    def __init__(self):
        self.radius = 10
        self.color = 'blue'

    def change_color(self, color):
        self.color = color

    # insert missing decorator here
    @classmethod
    def create_red_circle(cls):
        cls = Circle()
        cls.radius = 10
        cls.color = 'red'
        return cls

blue_circle = Circle()
red_circle = Circle.create_red_circle()

In [70]:
def multiples_of(number):
    for i in range(1, 6):
        yield i * number

In [71]:
multiples_of(number)

<generator object multiples_of at 0x00000276E46D3E58>

In [74]:
import pandas as pd
  
# dictionary of lists
dict = {'name':["aparna", "pankaj", "sudhir", "Geeku"],
        'degree': ["MBA", "BCA", "M.Tech", "MBA"],
        'score':[90, 40, 80, 98]}
  
df = pd.DataFrame(dict, index = [True, False, True, False])
  
df

name  degree  score
True   aparna     MBA     90
False  pankaj     BCA     40
True   sudhir  M.Tech     80
False   Geeku     MBA     98

In [75]:
# importing pandas as pd
import pandas as pd
  
# dictionary of lists
dict = {'name':["aparna", "pankaj", "sudhir", "Geeku"],
        'degree': ["MBA", "BCA", "M.Tech", "MBA"],
        'score':[90, 40, 80, 98]}
 
# creating a dataframe with boolean index
df = pd.DataFrame(dict, index = [True, False, True, False])
 
# accessing a dataframe using .loc[] function
df.loc[True]

name  degree  score
True  aparna     MBA     90
True  sudhir  M.Tech     80

In [76]:
df.iloc[1]

name      pankaj
degree       BCA
score         40
dtype: object

In [77]:
df.ix[True]

C:\Users\foram.r.solanki\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


name  degree  score
True  aparna     MBA     90
True  sudhir  M.Tech     80

In [78]:
df.ix[1]

C:\Users\foram.r.solanki\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


name      pankaj
degree       BCA
score         40
dtype: object

In [79]:
df[[True, False, True, False]]

name  degree  score
True  aparna     MBA     90
True  sudhir  M.Tech     80

In [82]:
import pandas as pd
  
# dictionary of lists
dict = {'name':["aparna", "pankaj", "sudhir", "Geeku"],
        'degree': ["BCA", "BCA", "M.Tech", "BCA"],
        'score':[90, 40, 80, 98]}
  
 
df = pd.DataFrame(dict, index = [0, 1, 2, 3])
 
mask = df.index == 0
 
df[mask]

name degree  score
0  aparna    BCA     90

In [85]:
import pandas as pd
  
# Define a dictionary containing employee data
data1 = {'Name':['Jai', 'Anuj', 'Jai', 'Princi',
                 'Gaurav', 'Anuj', 'Princi', 'Abhi'],
        'Age':[27, 24, 22, 32,
               33, 36, 27, 32],
        'Address':['Nagpur', 'Kanpur', 'Allahabad', 'Kannuaj',
                   'Jaunpur', 'Kanpur', 'Allahabad', 'Aligarh'],
        'Qualification':['Msc', 'MA', 'MCA', 'Phd',
                         'B.Tech', 'B.com', 'Msc', 'MA']}
    
  
# Convert the dictionary into DataFrame 
df = pd.DataFrame(data1)
  
df

Name  Age    Address Qualification
0     Jai   27     Nagpur           Msc
1    Anuj   24     Kanpur            MA
2     Jai   22  Allahabad           MCA
3  Princi   32    Kannuaj           Phd
4  Gaurav   33    Jaunpur        B.Tech
5    Anuj   36     Kanpur         B.com
6  Princi   27  Allahabad           Msc
7    Abhi   32    Aligarh            MA

In [89]:
df.groupby('Name')
df.groupby('Name').groups

{'Abhi': Int64Index([7], dtype='int64'),
 'Anuj': Int64Index([1, 5], dtype='int64'),
 'Gaurav': Int64Index([4], dtype='int64'),
 'Jai': Int64Index([0, 2], dtype='int64'),
 'Princi': Int64Index([3, 6], dtype='int64')}

In [90]:
gk = df.groupby('Name')
   
# Let's print the first entries
# in all the groups formed.
gk.first()

Age  Address Qualification
Name                              
Abhi     32  Aligarh            MA
Anuj     24   Kanpur            MA
Gaurav   33  Jaunpur        B.Tech
Jai      27   Nagpur           Msc
Princi   32  Kannuaj           Phd

In [91]:

# importing pandas module
import pandas as pd
  
# Define a dictionary containing employee data
data1 = {'Name':['Jai', 'Anuj', 'Jai', 'Princi',
                 'Gaurav', 'Anuj', 'Princi', 'Abhi'],
        'Age':[27, 24, 22, 32,
               33, 36, 27, 32],
        'Address':['Nagpur', 'Kanpur', 'Allahabad', 'Kannuaj',
                   'Jaunpur', 'Kanpur', 'Allahabad', 'Aligarh'],
        'Qualification':['Msc', 'MA', 'MCA', 'Phd',
                         'B.Tech', 'B.com', 'Msc', 'MA']}
    
  
# Convert the dictionary into DataFrame 
df = pd.DataFrame(data1)
  
print(df)

     Name  Age    Address Qualification
0     Jai   27     Nagpur           Msc
1    Anuj   24     Kanpur            MA
2     Jai   22  Allahabad           MCA
3  Princi   32    Kannuaj           Phd
4  Gaurav   33    Jaunpur        B.Tech
5    Anuj   36     Kanpur         B.com
6  Princi   27  Allahabad           Msc
7    Abhi   32    Aligarh            MA


In [93]:
# Using multiple keys in
# groupby() function
df.groupby(['Name', 'Qualification'])
 
df.groupby(['Name', 'Qualification']).groups

{('Abhi', 'MA'): Int64Index([7], dtype='int64'),
 ('Anuj', 'B.com'): Int64Index([5], dtype='int64'),
 ('Anuj', 'MA'): Int64Index([1], dtype='int64'),
 ('Gaurav', 'B.Tech'): Int64Index([4], dtype='int64'),
 ('Jai', 'MCA'): Int64Index([2], dtype='int64'),
 ('Jai', 'Msc'): Int64Index([0], dtype='int64'),
 ('Princi', 'Msc'): Int64Index([6], dtype='int64'),
 ('Princi', 'Phd'): Int64Index([3], dtype='int64')}

In [95]:
import pandas as pd
  
# Define a dictionary containing employee data
data1 = {'Name':['Jai', 'Anuj', 'Jai', 'Princi',
                 'Gaurav', 'Anuj', 'Princi', 'Abhi'],
        'Age':[27, 24, 22, 32,
               33, 36, 27, 32], }
    
  
# Convert the dictionary into DataFrame 
df = pd.DataFrame(data1)
  
df

Name  Age
0     Jai   27
1    Anuj   24
2     Jai   22
3  Princi   32
4  Gaurav   33
5    Anuj   36
6  Princi   27
7    Abhi   32

In [97]:
# using groupby function
# without using sort
 
df.groupby(['Name']).sum()

Age
Name       
Abhi     32
Anuj     60
Gaurav   33
Jai      49
Princi   59

In [98]:
# using groupby function
# with sort
 
df.groupby(['Name'], sort = False).sum()

Age
Name       
Jai      49
Anuj     60
Princi   59
Gaurav   33
Abhi     32

In [103]:

# importing pandas module
import pandas as pd
  
# Define a dictionary containing employee data
data1 = {'Name':['Jai', 'Anuj', 'Jai', 'Princi',
                 'Gaurav', 'Anuj', 'Princi', 'Abhi'],
        'Age':[27, 24, 22, 32,
               33, 36, 27, 32],
        'Address':['Nagpur', 'Kanpur', 'Allahabad', 'Kannuaj',
                   'Jaunpur', 'Kanpur', 'Allahabad', 'Aligarh'],
        'Qualification':['Msc', 'MA', 'MCA', 'Phd',
                         'B.Tech', 'B.com', 'Msc', 'MA']}
    
  
# Convert the dictionary into DataFrame 
df = pd.DataFrame(data1)
df

Name  Age    Address Qualification
0     Jai   27     Nagpur           Msc
1    Anuj   24     Kanpur            MA
2     Jai   22  Allahabad           MCA
3  Princi   32    Kannuaj           Phd
4  Gaurav   33    Jaunpur        B.Tech
5    Anuj   36     Kanpur         B.com
6  Princi   27  Allahabad           Msc
7    Abhi   32    Aligarh            MA

In [104]:

# using keys for grouping
# data
 
df.groupby('Name').groups

{'Abhi': Int64Index([7], dtype='int64'),
 'Anuj': Int64Index([1, 5], dtype='int64'),
 'Gaurav': Int64Index([4], dtype='int64'),
 'Jai': Int64Index([0, 2], dtype='int64'),
 'Princi': Int64Index([3, 6], dtype='int64')}

In [107]:

# importing pandas module
import pandas as pd
  
# Define a dictionary containing employee data
data1 = {'Name':['Jai', 'Anuj', 'Jai', 'Princi',
                 'Gaurav', 'Anuj', 'Princi', 'Abhi'],
        'Age':[27, 24, 22, 32,
               33, 36, 27, 32],
        'Address':['Nagpur', 'Kanpur', 'Allahabad', 'Kannuaj',
                   'Jaunpur', 'Kanpur', 'Allahabad', 'Aligarh'],
        'Qualification':['Msc', 'MA', 'MCA', 'Phd',
                         'B.Tech', 'B.com', 'Msc', 'MA']}
    
  
# Convert the dictionary into DataFrame 
df = pd.DataFrame(data1)
  
df

Name  Age    Address Qualification
0     Jai   27     Nagpur           Msc
1    Anuj   24     Kanpur            MA
2     Jai   22  Allahabad           MCA
3  Princi   32    Kannuaj           Phd
4  Gaurav   33    Jaunpur        B.Tech
5    Anuj   36     Kanpur         B.com
6  Princi   27  Allahabad           Msc
7    Abhi   32    Aligarh            MA

In [108]:

# iterating an element
# of group
 
grp = df.groupby('Name')
for name, group in grp:
    print(name)
    print(group)
    print()

Abhi
   Name  Age  Address Qualification
7  Abhi   32  Aligarh            MA

Anuj
   Name  Age Address Qualification
1  Anuj   24  Kanpur            MA
5  Anuj   36  Kanpur         B.com

Gaurav
     Name  Age  Address Qualification
4  Gaurav   33  Jaunpur        B.Tech

Jai
  Name  Age    Address Qualification
0  Jai   27     Nagpur           Msc
2  Jai   22  Allahabad           MCA

Princi
     Name  Age    Address Qualification
3  Princi   32    Kannuaj           Phd
6  Princi   27  Allahabad           Msc



In [111]:
t = (1,5,"Disco", ("Python", "Java"))
type(t[3])

tuple

In [112]:
 dir()

['Circle',
 'GameProtagonist',
 'In',
 'L1',
 'L2',
 'Out',
 'S',
 'Str',
 '_',
 '_100',
 '_103',
 '_104',
 '_105',
 '_107',
 '_109',
 '_110',
 '_111',
 '_14',
 '_16',
 '_2',
 '_20',
 '_3',
 '_33',
 '_36',
 '_39',
 '_4',
 '_5',
 '_51',
 '_52',
 '_54',
 '_59',
 '_6',
 '_60',
 '_61',
 '_63',
 '_64',
 '_65',
 '_7',
 '_71',
 '_74',
 '_75',
 '_76',
 '_77',
 '_78',
 '_79',
 '_80',
 '_81',
 '_82',
 '_85',
 '_87',
 '_88',
 '_89',
 '_90',
 '_93',
 '_95',
 '_96',
 '_97',
 '_98',
 '__',
 '___',
 '__builtin__',
 '__builtins__',
 '__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_dh',
 '_i',
 '_i1',
 '_i10',
 '_i100',
 '_i101',
 '_i102',
 '_i103',
 '_i104',
 '_i105',
 '_i106',
 '_i107',
 '_i108',
 '_i109',
 '_i11',
 '_i110',
 '_i111',
 '_i112',
 '_i12',
 '_i13',
 '_i14',
 '_i15',
 '_i16',
 '_i17',
 '_i18',
 '_i19',
 '_i2',
 '_i20',
 '_i21',
 '_i22',
 '_i23',
 '_i24',
 '_i25',
 '_i26',
 '_i27',
 '_i28',
 '_i29',
 '_i3',
 '_i30',
 '_i31',
 '_i32',
 '_i33',
 '_i34',
 '_i35',
 '_i36

In [114]:
L1 = [10,20,30,24,18]
L2 = [8,14,15,20,10]
L3 = [0,1,2,3,4]

for i in L3:
    L3[i] = L1[i] - L2[i]
    
L3

[2, 6, 15, 4, 8]

In [115]:
list_numbers = (1,2,3,4)
sample_map = map(lambda x: x*2, list_numbers)
list(sample_map)

[2, 4, 6, 8]

In [116]:
from functools import reduce
list_1 = ['harshit','aman']
reduce(lambda x,y: x + y,list_1)

'harshitaman'